### PR TITLE
fix: Fix unsafe std::optional accesses

### LINKED ISSUE #755: Bug: some instances of std::optional are dereferenced without checking if they hold a value
While investigating what happens with bound parameters when their covariance matrices are std::nullopt, we have found instances where the optional values are unconditionally dereferenced which is undefined behavior if the optional does not hold anything.

See attached file for a list of potential instances of this problem in the case of BoundTrackParameter::covariance(). Note that some of those may be legitimate if the optional is verified beforehand, we will have to check. Also, there may be other instances of this kind of pattern elsewhere, this should be verified as well.

See the following mattermost thread: https://mattermost.web.cern.ch/acts/pl/jap67ecwmir43pbyh68nn5jwjy

### PR DESCRIPTION
**Pull request by gagnonlg**
This PR fixes unsafe instances of std::optional reported in #755 .

I attempted to pick the solution that made the most sense at each specific instances, so there's a mix of has_value() guards, checked access with .value(), or returning error codes.

Fixes #755 

Pinging @paulgessinger

### CODE REVIEW

**Comment by gagnonlg** [comment]
@paulgessinger can you review this one?

**Comment by paulgessinger** [comment]
Hm, one thing I see is that `.value()` and `.has_value()` have runtime overhead. `operator*()` doesn't, and results in UB if `!opt.has_value()`.

Would it be reasonable to add assertions for `DEBUG`, and run the risk of UB for `NDEBUG`.

File: Core/include/Acts/Vertexing/GaussianTrackDensity.ipp [hunk 1]
```diff
@@ -78,6 +78,9 @@ void Acts::GaussianTrackDensity<input_track_t>::addTracks(
     const double d0 = boundParams.parameters()[BoundIndices::eBoundLoc0];
     const double z0 = boundParams.parameters()[BoundIndices::eBoundLoc1];
     // Get track covariance
+    if (not boundParams.covariance().has_value()) {
+      continue;
```
Comment (paulgessinger) [inline Core/include/Acts/Vertexing/GaussianTrackDensity.ipp:82]: I would assume this is an error?

Comment (gagnonlg) [reply Core/include/Acts/Vertexing/GaussianTrackDensity.ipp:82]: The problem here is that this function returns `void` and not e.g. a  `Result` type. If we want to signal an error here we can just throw and exception (e.g. by using `.value`)?

**Comment by gagnonlg** [comment]
> Hm, one thing I see is that `.value()` and `.has_value()` have runtime overhead. `operator*()` doesn't, and results in UB if `!opt.has_value()`.
> 
> Would it be reasonable to add assertions for `DEBUG`, and run the risk of UB for `NDEBUG`.

I think we can afford the slight overhead. e.g. I check with godbolt & at least a recent gcc optimizes the `has_value()` call to just two instructions (cmp + jump). Same for `value()`; cmp + jmp, and the more costly exception branch is only entered if there's no value. So I'd be very surprised if the  overhead is noticeable.

**Comment by paulgessinger** [comment]
@gagnonlg you're probably right.

File: Core/include/Acts/Vertexing/GaussianTrackDensity.ipp (hunk 1 above)
Comment (paulgessinger) [reply Core/include/Acts/Vertexing/GaussianTrackDensity.ipp:82]: Do we have a logger here? I think that would help make the error more traceable. Then we could either do the exception or just an `std::terminate`. In a follow-up PR, we could switch the return to `Result<void>`. What do you think?

File: Examples/Io/Root/src/RootTrajectoryParametersWriter.cpp [hunk 2]
```diff
@@ -201,21 +201,26 @@ ActsExamples::ProcessCode ActsExamples::RootTrajectoryParametersWriter::writeT(
...
+              sqrt(covariance(Acts::eBoundLoc0, Acts::eBoundLoc0));
+          m_err_eLOC1_fit =
+              sqrt(covariance(Acts::eBoundLoc1, Acts::eBoundLoc1));
+          m_err_ePHI_fit = sqrt(covariance(Acts::eBoundPhi, Acts::eBoundPhi));
+          m_err_eTHETA_fit =
+              sqrt(covariance(Acts::eBoundTheta, Acts::eBoundTheta));
+          m_err_eQOP_fit =
+              sqrt(covariance(Acts::eBoundQOverP, Acts::eBoundQOverP));
+          m_err_eT_fit = sqrt(covariance(Acts::eBoundTime, Acts::eBoundTime));
+        }
```
[the remark this answers is not in the document]
Comment (gagnonlg) [reply Examples/Io/Root/src/RootTrajectoryParametersWriter.cpp]: Done

File: Core/include/Acts/Vertexing/GaussianTrackDensity.ipp (hunk 1 above)
Comment (gagnonlg) [reply Core/include/Acts/Vertexing/GaussianTrackDensity.ipp]: The function returns `void`, so it was trivial to change it to `Result<void>` and use `VertexingError::NoCovariance` to signal the error. This function is called in a single other function which already returns a special value if its computation is invalid, so I did the same thing for the `addTracks`. Let me know if that works!

**Review by paulgessinger — APPROVED** [review]
I guess this is fine.

**Pull request merged**